pkg/logql/log/parser.go

```diff
@@ -259,3 +260,49 @@ func (l *LogfmtParser) Process(line []byte, lbs *LabelsBuilder) ([]byte, bool) {
 }
 
 func (l *LogfmtParser) RequiredLabelNames() []string { return []string{} }
+
+type JSONExpressionParser struct {
+	expressions map[string][]interface{}
+}
+
+func NewJSONExpressionParser(expressions []JSONExpression) (*JSONExpressionParser, error) {
+	var paths = make(map[string][]interface{})
+
+	for _, exp := range expressions {
+		path, err := jsonexpr.Parse(exp.Expression, false)
+		if err != nil {
+			return nil, fmt.Errorf("cannot parse expression [%s]: %w", exp.Expression, err)
+		}
+
+		paths[exp.Identifier] = path
```

Comment:
I think you need to validate that identifier is a valid prometheus label name. I think other parser does that somewhere ?

Reply:
Yeah, the `exp.Identifier` here uses the `IDENTIFIER` type in the grammar, so it _should_ be valid according to that - same as `label_format` etc

Reply:
👉  https://github.com/grafana/loki/blob/master/pkg/logql/log/parser.go#L191

Reply:
Ah, interesting - thanks. Seems that `label_format` doesn't use this at all - probably because invalid identifiers will not be acceptable in the LogQL definition, whereas regex identifiers are defined as strings.
Should I explicitly add this check?

Reply:
> probably because invalid identifiers will not be acceptable in the LogQL grammar,

Not sure about that, may be the label_format is allowing thing it should not :)

Reply:
I don't think so:
![image](https://user-images.githubusercontent.com/373762/107368088-c9786680-6ae8-11eb-9aae-efa9eb9ac15c.png)
![image](https://user-images.githubusercontent.com/373762/107368174-e3b24480-6ae8-11eb-9072-bc48f2ad4707.png)

In any case, adding the validation won't make a huge difference, added :+1: 

No impact on benchmarks:

```bash
$ go test -run=NONE -bench='JSON.*Parser$' -benchtime 5s -count 3 ./pkg/logql/log
goos: linux
goarch: amd64
pkg: github.com/grafana/loki/pkg/logql/log
BenchmarkJSONParser-8                    3039807              2167 ns/op             992 B/op         14 allocs/op
BenchmarkJSONParser-8                    3093699              1942 ns/op             992 B/op         14 allocs/op
BenchmarkJSONParser-8                    3026457              1990 ns/op             992 B/op         14 allocs/op
BenchmarkJSONExpressionParser-8          3197628              1804 ns/op             936 B/op         16 allocs/op
BenchmarkJSONExpressionParser-8          3327074              1923 ns/op             936 B/op         16 allocs/op
BenchmarkJSONExpressionParser-8          3061735              1948 ns/op             936 B/op         16 allocs/op
PASS
ok      github.com/grafana/loki/pkg/logql/log   48.425s```

Reply:
I could be wrong but `#foo` might got accepted by the lexer/parser but is not valid.

Reply:
Anything that follows a `#` is considered a comment, since my [PR](https://github.com/grafana/loki/pull/3126)